src/migrate.ts

```diff
@@ -0,0 +1,71 @@
+"use strict"
+// chel migrate --from fs --to sqlite --out ./database.db ./data
+
+import { colors, flags, path } from './deps.ts'
+import { exit, isDir, isFile, isHashKey, isValidKey } from './utils.ts'
+
+const backends = {
+  fs: await import('./database-fs.ts'),
+  sqlite: await import('./database-sqlite.ts')
+}
+
+export async function migrate(args: string[]) {
+  const parsedArgs = flags.parse(args)
+
+  const { from, to, out, _ = ['.'] } = parsedArgs
+  const src = path.resolve(_[0])
+
+  if (!from) exit('missing argument: --from')
+  if (!to) exit('missing argument: --to')
+  if (!out) exit('missing argument: --out')
+
+  if (!backends[from]) exit(`unknown storage backend: "${from}"`)
+  if (!backends[to]) exit(`unknown storage backend: "${to}"`)
+  if (from === to) exit('arguments --from and --to must be different')
+
+  if (isDir(src)) {
+    if (from === 'sqlite') exit(`not a database file: "${src}"`)
+  } else if (isFile(src)) {
+    if (from === 'fs') exit(`not a directory: "${src}"`)
+  } else {
+    exit(`not found: "${src}"`)
+  }
+
+  if (isDir(out)) {
+    if (to === 'sqlite') exit(`argument --out is a directory: "${out}"`)
+  } else if (isFile(out)) {
+    if (to === 'fs') exit(`argument --out is a file: "${out}"`)
+  } else if (out.endsWith('./')) {
+    if (to === 'sqlite') exit(`argument --out ends with a slash: "${out}"`)
+  }
+
+  try {
+    backends[from].initStorage(from === 'fs' ? {dirname: src} : {dirname: path.dirname(src), filename: path.basename(src)})
+  } catch (error) {
+    exit(`could not init storage backend at "${src}" to migrate from: ${error.message}`)
+  }
+  try {
+    backends[to].initStorage(to === 'fs' ? {dirname: out} : {dirname: path.dirname(out), filename: path.basename(out)})
+  } catch (error) {
+    exit(`could not init storage backend to migrate to: ${error.message}`)
+  }
+
+  const keys = (await backends[from].readKeys()).filter(isValidKey)
```

Comment:
Have you considered using some sort of streaming API for `readKeys` so that we can handle databases with hundreds of thousands of keys without massive memory usage?

Reply:
I did, but at first I didn't find one in the sqlite API for Deno. I just checked the docs again now that I have a bit more time, and `https://deno.land/x/sqlite@v3.7.1/mod.ts?s=PreparedQuery&p=prototype.iter` looks promising!